### DISCUSSION TITLE
Which new note type should we implement next?

### DISCUSSION THREAD

**Discussion by adoriandoran** [TriliumNext/Trilium#5986] [General]
What new note types would you like to see implemented in TrilliumNext?

**Comment by adoriandoran** [comment]
#132 

I propose adding a mindmap note type. While TrilliumNext currently supports mindmaps through Mermaid Diagrams, it's not very user-friendly or comfortable for brainstorming sessions. A visual editor would be ideal.

**Reply by meichthys** [reply]
The canvas note type may get you close to what you want, but i agree that a mind-mapping note type could be useful 👍

**Reply by adoriandoran** [reply]
> The canvas note type may get you close to what you want, but i agree that a mind-mapping note type could be useful 👍

Indeed, Canvas can also create mind maps. However, I prefer when the software automatically manages the layout, so I can focus solely on the ideas rather than spending time arranging the drawing elements.

**Reply by bitkata** [reply]
A Kanba Board would be nice. The extension by [CyrilLeblanc](https://github.com/CyrilLeblanc) doesn't seem to work.

**Reply by q7w7e7r** [reply]
Yes, kanban would be great to add. But I'm still not happy with the functionality of todo and I don't know how to improve it or what to replace it with.

**Comment by adoriandoran** [comment]
#90

A spreadsheet could also be useful for notes that require automatic calculations and sorting.

**Reply by meichthys** [reply]
Related: https://github.com/TriliumNext/Notes/issues/90

**Reply by rom1dep** [reply]
I suspect something like https://github.com/dream-num/univer would deliver part or all of that

**Reply by CobriMediaJulien** [reply]
+1 this would be huge

**Comment by LarsaSara** [comment]
Passwords management.

**Reply by meichthys** [reply]
Just thinking out loud here, but you should be able to use [promoted attributes](https://triliumnext.github.io/Docs/Wiki/promoted-attributes.html) to partly/mostly achieve this:

For example:
![image](https://github.com/user-attachments/assets/9de4b8c6-e913-4ea7-b32c-aa3d55d09956)

Attributes used:
![image](https://github.com/user-attachments/assets/bd84952d-1082-4825-a829-b02fa42a7ab3)

This along with being able to password protect a credential note and/or parent note (that contains credentials) could probably get you close to a 'credential type note'.

To use this, you would probably also want to make a 'base' credential note and convert it into a [template note](https://triliumnext.github.io/Docs/Wiki/template.html):
![image](https://github.com/user-attachments/assets/6e7b0d7f-ad6a-4bdf-99b0-d0b8fdf8025c)

This would allow you to create new 'credential notes' easily like:
![image](https://github.com/user-attachments/assets/813dc895-0dd2-466b-9b3b-318a85e2ac69)

**Reply by eliandoran** [reply]
Seems feasible, we should probably add a password field type, to hide the password. 🤔 
However I wonder if it would be useful under any circumstances.

**Reply by meichthys** [reply]
Yeah, maybe not necessary since we can password protect notes already. However, it would be nice if we could password protect notes using a different credential than the users's login credential.

It's an intresting idea nonetheless. Just not sure how practical it would be since the featureset of a 'credential' note would never compare to a full-on password manager.

**Comment by Alumniminium** [comment]
https://obsidian.rocks/getting-started-with-zettelkasten-in-obsidian/

This keeps coming up when I tell people to switch to trilium and they already use Obsidian.

**Reply by eliandoran** [reply]
@Alumniminium , is this something that can be implemented as a note type? 🤔 
To me it seems more like an personal organization thingie that can be achieved using the note hierarchy.

**Reply by Alumniminium** [reply]
to be honest i never understood what this is :smiling_face_with_tear:

**Reply by EventualDankDog** [reply]
I also think it's more about a way to organize your notes. Can be done perfectly with internal links and relations

**Reply by CobriMediaJulien** [reply]
It´s jsut a way to organize your notes and zadam pretty followed this tecnique designign trilium architecture

**Reply by 0xbismarck** [reply]
Been reading through the link you provided above, and a few of the links within the article. The only thing that comes to mind is a [Map of Content](https://obsidian.rocks/quick-tip-quickly-organize-notes-in-obsidian/). As @EventualDankDog said, this probably could be done with existing internal links and link previews. But figured I'd suggest it in case it resonates.  To build an MOC now is more manual in Trilium, and maybe something that could support dynamic MOC changes might interest people.

As an aside: I do find it interesting that BASB was one of the additional resources. I been wondering if PARA and Zettlekasten could complement each-other or not.

**Comment by kangaroolab** [comment]
How about a notebook type? It will be handy if such is provided to do both text and run code (with extension to use different engines) on the same note.

**Comment by tevenfeng** [comment]
1. Some sort of video type? For example maybe I would like to include/insert a youtube embed video into my note.
2. Collection and Set function like Anytype, maybe just based on the label system.

**Reply by zerebos** [reply]
> Collection and Set function like Anytype, maybe just based on the label system.

Being able to do arbitary collections, sets, and _objects_ would also help Trilium rival Anytype, capacities, and other similar PKMs.

**Comment by abstractionmage** [comment]
I'd love to see a World Map note similar to [Obsidian's Map View community plugin](https://github.com/esm7/obsidian-map-view?tab=readme-ov-file#intro) so I can finally place all my travel notes on a world map and get a geographical overview of all the nice places I've already been to.

Maybe we could use [Leaflet](https://leafletjs.com) for that, it seems to be a pretty popular javascript library for exactly this kind of functionality.

**Reply by masseym** [reply]
This sounds very cool

**Comment by herrkami** [comment]
**A text-based 2D/3D CAD note.** 

The benefits would be shiny and plenty imho:

- precise, reusable, and scalable sketching 
- STL-support would allow saving, viewing and, perhaps, even editing 3D parts as notes
- combined with Code Notes, it would create a powerful environment for developing entire projects anywhere and from everywhere
- it would close an important gap for people who use Trilium to sketch and develop real-world projects (e.g. furniture, art, mechanical thingies, electronic gadgets, etc.)

And I guess, this would be far easier and faster to implement then it might appear at first glance. 

For example, there's [the web version of JSCAD](https://github.com/jscad/OpenJSCAD.org/tree/master/packages/web). It's easy to customize, runs in the frontend, and would integrate well with the existing interface for frontend JS code notes. It even uses CodeMirror already. Feature- and compatibility-wise it's likely more than capable to turn Trilium into a fairly powerful CAD suite. The CAD-language is easy to learn and practical. 

A better CAD-language would be [build123d](https://build123d.readthedocs.io/en/latest/) for example. Unfortunately, this would be far harder and costly to implement, as it requires Python, and hence, an entire (safe and secure!) backend environment for Python. Source notes would then have to be built within this environment and sent back as renderings to a frontend renderer, e. g. based on [three-cad-viewer](https://github.com/bernhard-42/three-cad-viewer) or so. I'd strongly vote against this option for many (obvious) reasons. 

JSCAD would be great though. I'm currently trying to implement it into Trilium via Render Notes and Templates. If that works out, it could even be implemented and shipped as a widget or so. However, for better usability, maintenance, and efficiency, a dedicated note type would be superior imo.  

Cheers, Kami

**Comment by diyoyo** [comment]
I'd say a RelationMap that is built only from a selected set of existing notes, instead of having to create new ones to create the network.

Also, my dream would be to be able to add metadata to edges:
If `(A:Person)-[workedAt]->(B:Company)`, I'd like to provide information on the **"workedAt"** relationship, like which year, what job, etc.

**Comment by diyoyo** [comment]
Also, some other thoughts:
- maybe a SQLite db could be interesting to have, as a note.
- a pic slideshow

**Comment by yu-jingrui** [comment]
Not a new note type but a new way to display child notes:
currently the child notes are displayed as fixed-size boxes under the parent note.
What I would like is to **display the child notes one by one horizontally under the main content of the parent note**, with the possibility to show only the first few lines or the whole of the child note.

The use case is to create a main post, under which I constantly add new contents to the same topic, without having to tamper with the main post, but at the same time have an overview of what I have done under this topic. This will enable keeping track of (for example) a journal of constent developement in one topic.

**Reply by eliandoran** [reply]
Hi, @yu-jingrui .

Your description seems to correspond to a feature that is already present in the application, can you confirm? More specifically, it's the Book functionality which looks like this:

<img width="1092" alt="image" src="https://github.com/user-attachments/assets/a667994f-349f-4ead-be71-9af0bb24985c">

You can configure it by selecting in the "Basic properties" ribbon tab, the note type of "Book" (usually this is done in an empty note as it will no longer display its text content, only child notes). And then on the "Book properties" tab you have the possibility to select "List" instead of "Grid".

If this is not adequate, let me know what you think could be improved and we can consider it.

**Reply by yu-jingrui** [reply]
Hi @eliandoran,
this is almost what I want, I need to read the doc more carefully.
There is but one problem that the contents of the parent note is not shown in book mode.
However, I found sth strage. I changed the view type to `list` in book mode, and changed the note back to text mode, and the list view of child notes remained. I ended up with a text type parent note that shows the child notes in list view. I found that the note now has the attribute `viewType=list`.  Also, although the content of the parent note does not show when it's in book mode, it is not lost when changed back to text mode.
My guess is there is no real difference between the two note types (book, text), only how to display. So it should be possible to enable the switch of grid/list view for text note.

**Reply by rom1dep** [reply]
> Your description seems to correspond to a feature that is already present in the application, can you confirm? More specifically, it's the Book functionality which looks like this

I have a problem with that, supposedly, how a note should look like should be a display/rendering property, not a "note type" thing. Is there a technical reason for the way it is?

**Reply by perfectra1n** [reply]
@rom1dep are you suggesting that the the "view" the Book note type provides (e.g. in the screenshot above) should be available to other note types as well? Such as beneath a Text Note?

**Reply by yu-jingrui** [reply]
@perfectra1n I find the "book" type has nothing fundamentally different from "text". The only real difference is that "book" has no own content text. I think there is no need to retain "book" type, and I agree with @rom1dep that it should be a display/rendering property.

**Reply by rom1dep** [reply]
> @rom1dep are you suggesting that the the "view" the Book note type provides (e.g. in the screenshot above) should be available to other note types as well? Such as beneath a Text Note?

I think so, yeah. My impression (but again, I could be wrong!) is that such a rendering option could be relevant to anything that has child notes.

**Reply by perfectra1n** [reply]
Cool, got it. It should be pretty easy to find in the code, if anyone wants to take a crack at it before Elian or I can - I'm assuming that it's exactly what's been proposed here. I'm sure there's some flag to check that "Book" has been selected and then it renders the "view" in the screenshot above.

**Reply by perfectra1n** [reply]
@rom1dep would you mind tossing this into an issue and assigning it to me? I wanna make sure I don’t miss anything, I’m curious to look into it

**Reply by rom1dep** [reply]
@perfectra1n  I'd love to, especially since there might be some good bits to include from https://github.com/orgs/TriliumNext/discussions/356#discussioncomment-11146681 , sadly I'll be travelling for few weeks and might have to pick that up in a while.

**Comment by kleutzinger** [comment]
Audio notes or some type of voice recording. It may or may not have to be a note type, it could also just be some object you embed inside a note or as some type of attatchment.

Another nice-to-have would be an automatic timestamped transcription (with click-to-jump on the text). Google Keep (or https://recorder.google.com/) are a nice reference for well-implemented voice notes.

**Reply by meichthys** [reply]
I can see the usefulness of this. It was recently requested on the original Trulium repo.

**Reply by perfectra1n** [reply]
Is anyone aware of any useful/quality self-hosted transcription service or tool? Might be cool to integrate with the voice recording as well.

**Reply by aaronsantiago** [reply]
https://github.com/xenova/whisper-web/tree/experimental-webgpu is an option, some self-hosted API-esque services listed here: https://github.com/sindresorhus/awesome-whisper?tab=readme-ov-file

**Comment by glaucon1984** [comment]
I have a suggestion that might be easy to implement: a "markdown" type note. 
This would be a note that has a split-screen view, one side is plain text, the other shows the rendered note. 
I say it might be easy to implement because there is already an approximation to this: 
https://github.com/zerebos/Trilium-MarkdownPreview
This type of note could make it easier for users coming from Joplin or Obsidian.

**Reply by Winson-Huang** [reply]
In fact, the editor in Trilium supports some markdown grammer, but the format in "import from / export to markdown" is not good.....

**Comment by EventualDankDog** [comment]
A musical note sheet

**Reply by meichthys** [reply]
Interesting idea. Do you have any samples or tools that implement something similar?

**Reply by rom1dep** [reply]
Apparently there was an active attempt at porting lilypond to the web, assuming this is what OP is talking about: https://www.weblily.net/

**Reply by EventualDankDog** [reply]
Unfortunately I don't have anything similar, it was just an idea. Sometimes I come up with ideas for melodies when I'm playing around on the piano and my synth but I'm to lazy to turn on the PC and record it. That's why I thought it might be handy

**Reply by pano9000** [reply]
I like this idea, but a question:
would this be about rendering the data as a musical note sheet only, or actually *entering* the data into a note sheet?

there would be a few libraries that can do the former with some "simplified" text based language, e.g.
https://www.abcjs.net/#what

**Reply by meichthys** [reply]
If the former was implemented, one could use the Include notes feature to display other types of notes inside of a regular text note.

**Reply by EventualDankDog** [reply]
> I like this idea, but a question:
> would this be about rendering the data as a musical note sheet only, or actually *entering* the data into a note sheet?
> 
> there would be a few libraries that can do the former with some "simplified" text based language, e.g.
> https://www.abcjs.net/#what

So originally I was thinking about the latter one, but abcjs kinda looks cool and since I'm writing down melodies in a similar way already, rendering it into actual notes would be a really nice feature for my use case.

**Comment by diyoyo** [comment]
A ChatGPT integrated discussion. It's such a pain to copy paste the conversations and keep a nice formatting. So why not provide a way to enter an API token and have a discussion already in Trilium ?

**Reply by perfectra1n** [reply]
You can give [this](https://github.com/soulsands/trilium-chat) a shot if you're interested @diyoyo :)

**Comment by brianvanderburg2** [comment]
Engineer Calculator note type.

While this can probably already be achieved with Render Note/etc a dedicated note type could make adding utility calculators easier.  One can define a series of fields for input, their unit types, and a calculate function.  Some calculators may be simple with only one output from one or more inputs, but it should also be possible to define more complex calculators that may take different inputs and have multiple outputs,

Unit types define for instance length or time and a set of conversions in a array ("ft": 1.0, "yd": 3.0, "in": 1.0 / 12.0, "cm": 1.0 / 30.5)  A note can define field types and they would displayed to the user. The user can select different conversion from the same type as input.  In the code, field1.getValue("ft"), or something like that, would automatically convert the input values for the calculator function based on the user's selection and the unit type of the field,  The code can set output via field2.setValue("s", 5.0) and the field would automatically reflect the user unit selection.

Attachments can be added for visual representations to show what the fields represent if needed.  There should exist some built-in unit types, "unit", "time", "length", "mass", "weight", "speed", etc.  It should also be possible to extend the unit types with unit-type notes, so calculator notes don't have to define the same set of unit types over and over.  Unit conversions can also be extended:

builtin_length = [["ft", 1.0], ["yd", 3.0], ["in", 1.0 / 12.0], ["cm": 1.0 / 30.5]]

conversion extension note:

yd = conversions["length"].get("yd")
extra_length  = [["parsec", yd * 3.374563e16]]
conversions["length"].extend(extra_length)

**Reply by brianvanderburg2** [reply]
I've been working on a widget that implements something like this, at very rough early stages

https://gist.github.com/brianvanderburg2/a23384b73be410685aa4575d7554b78f

**Comment by asamarco** [comment]
Forms, a note format for structured data. Same idea of promoted attributes, key value pairs but in the main text of the note. Possibly with additional field types including dropdown menus/radio buttons/checkboxes. Effectively making a note template the definition of a database table.

**Reply by meichthys** [reply]
Neat idea. I like it 👍

**Reply by rom1dep** [reply]
Are you proposing to have a "Database" type of note, somewhat similar to Siyuan or Notion? If so, I would find the approach to clash with the current paradigm that the whole notebook (i.e. "Document") __is__ the database.

With the tools we already have at our disposal (attributes, templates, inheritance and cloning), we can do quite a lot, like define re-usable "Typed" records (e.g. instances of "Persons" created from a Template note with owned attributes), that are extensible (e.g. "Colleagues" can be extensions of "Persons" with refined attributes) and composable (permitting "sum typing", for instance, you can define "Colleagues" that are "Friends").

Perhaps if we could formalize all that into a sound type system/algebra, we could build a decent UI to manage notes' metadata at scale (something like https://github.com/mabeyj/trilium-collection-views#table-view with bells and whistles).

**Reply by asamarco** [reply]
> Are you proposing to have a "Database" type of note, somewhat similar to Siyuan or Notion? If so, I would find the approach to clash with the current paradigm that the whole notebook (i.e. "Document") **is** the database.
> 
AFAIK the only software that can do what I think is [treeline](https://github.com/doug-101/TreeLine). See https://github.com/asamarco/Cleanroom-ELN as an example of heavily object-based notebook

Neither Notion or Siyuan have this capability. The same paradigm of document is database is followed with my example.

Trillium is really close to what I would use because promoted attributes are already a barebone implementation of a more structured form.

**Reply by meichthys** [reply]
It would be nice just to be able to place the promoted attributes directly inline with the note content.

**Reply by eliandoran** [reply]
> It would be nice just to be able to place the promoted attributes directly inline with the note content.

@meichthys , that's exactly what I'm trying to do with https://github.com/TriliumNext/Notes/pull/538 .

I've removed the promoted attributes and put them below the title and the content. The only problem is that in that PR I've also tried to make the title of the page also in-line, similar to how more _modern_ editors such as Notion do it, but it introduces a few layout problems.

**Reply by rom1dep** [reply]
Treeline is fascinating. I only spent very little time messing around with it but I am still convinced that Trilium is more capable in terms of data modeling. Anyhow, we all agree that the mean to manage notes as data in Trilium is via promoted/inherited attributes, and there should be means to render those as tables with R/W capabilities. @asamarco Any opinion on the [table-view](https://github.com/mabeyj/trilium-collection-views#table-view) extension I linked previously?

**Reply by asamarco** [reply]
> @asamarco Any opinion on the [table-view](https://github.com/mabeyj/trilium-collection-views#table-view) extension I linked previously?

I installed it, it could be nice to visualize homogeneus sets of data like the example of books but it is far to what I would use to help me organize hyerarchies of objects. What is missing is the capability to represent a tree of different objects, whose depth, content, and even parent type can be different.  I acknowledge that the Cleanroom-ELN template I linked above is a bit too niche and technical but it shows exactly how versatile trees and custom object templates can become. Trillium has already a hierarchy of notes, so by using templates and improved attributes might as well provide this way of data management which is lacking in all the other note-taking software right now.

**Reply by rom1dep** [reply]
> it could be nice to visualize homogeneus sets of data like the example of books but it is far to what I would use to help me organize hyerarchies of objects.

I suppose Trilium encourages "homogeneous sets of data" in how it makes it much easier to handle large amount of typed notes through Inheritance (much more so than through Templates, which have stark downsides, like the absence of automatic attributes updates whenever the master template changes).  
An approach I use a lot is to build a "Library" note with subtrees (e.g. "People", "Places", …) containing a bunch of notes of a same type that I then clone or link in context. In that scenario, the [table-view](https://github.com/mabeyj/trilium-collection-views#table-view) approach really shines, because a bi-directional editor/widget could be hooked-up at the head of one of those collections and enable quick and convenient grid edition of all notes of the same type (i.e. the list of all "People", "Places", … with all their common attributes as columns).

> What is missing is the capability to represent a tree of different objects, whose depth, content, and even parent type can be different.

I suspect trilium's query engine is quite capable in this area: https://triliumnext.github.io/Docs/Wiki/search.html , and it should be a feature-goal IMO to tap into it like table-view does. But again, in practice, I wouldn't really know what to expect from a list of heterogeneous notes. Perhaps you could share a screenshot from Treeline showing how you like it?

**Reply by asamarco** [reply]
![RDT_20241104_2234268791224178132384294](https://github.com/user-attachments/assets/0ce02a22-8f61-4495-8a99-a452804d9e93)
![RDT_20241104_2234312615458955561667571](https://github.com/user-attachments/assets/a73905e8-c62e-4e12-8c25-3552926f5f66)
There could be easier examples but this is what I already have ready in treeline. This is how I keep track of my production flows. I have about 40 object types representing different fabrication processes, physical objects, containers, measurements... At top level there is a research sample, underneath the processes it withstand in the right order and possibly subprocesses if required. Processes can change, repeat, fork. It can become quite complicated and non linear (https://imgur.com/a/19KJvxt) so it definitely require a tree to keep track of the reality.

In any case I was just suggesting to have a note type that is able to represent nicely key:value pairs. I believe would be appreciated even if I could not replace treeline anyway.

**Reply by rom1dep** [reply]
I don't find this enumeration of key-values to be very convenient nor practical, but I can see how it's a lesser evil in your workflow. IMO, this attributes enumeration on the right should be a tree of expandable/collapsible nodes just like the nodes hierarchy it mirrors on the left. I suspect it would be trivial to implement over a refacto/generalization of the book note type (discussed above).

**Comment by RulentWave** [comment]
I don't know how difficult this would be, but a spreadsheet note type would be very useful. Perhaps there is a library to embed ODFs?

**Reply by rom1dep** [reply]
already discussed here: https://github.com/orgs/TriliumNext/discussions/356#discussioncomment-10362274

**Comment by pradt** [comment]
Something to model data would be good - ER diagram etc ? 

Maybe something similar to the functionality of the relationship diagram, but allow to do one to many, many to many etc... relationships.

**Reply by perfectra1n** [reply]
ERD Diagrams already work, using the Mermaid note type :)

**Comment by 0xbismarck** [comment]
Maybe a series of planner notes? Previously in OneNote, I've built my own version of the planner described here: https://www.youtube.com/watch?v=5YIqHMoslTk

I'm not sure how I'd start to flesh out something like this in Trilium.

**Reply by eliandoran** [reply]
@0xbismarck , could you give more details about your implementation? The video is interesting, but it's an hour long so I would appreciate an overview.

In addition, do note that we added the calendar view in the latest betas which might be of interest. Perhaps all we need is to expand it a little bit further, and that's where the feature requests should go in.

**Reply by 0xbismarck** [reply]
Agree it's long. 

For my OneNote workflow, I've setup four types of planner notes based on this. (Providing timestamps to show you layouts)

The annual planner - at the 14:50 mark.

The monthly dashboard - 32:28

Switching to this video - https://www.youtube.com/watch?v=JcZGH91trYU

Monthly dashboard - 3:33

Weekly Dashboard - 17:05

Daily Page - 26:01 

When planning my day in  OneNote, I go from Monthly → Weekly → Daily in order to build out my task list for the day. I've also got them linked to each other so I can jump between Daily to Weekly & Monthly as needed. 

Mainly I'm showing layout ideas. I've slowly tweaked it as I've been using it for about a year. The things that make it useful is I have the links to my active projects in there, so I can go straight to my notes on those project, I have a listing of the tasks I want to try and complete for the day, and I can quickly jot time things like what did, or something I want to remind myself for later. 

I haven't seen the calendar view yet. I'll have to check it out.

**Comment by aaronsantiago** [comment]
Something like https://glitch.com/ might be very cool. Left side panel for code, a right-side panel with rendered HTML/JS that has a clear entrypoint div for scripting from the left side. Optional properties for how it should be rendered when linked would be interesting, and properties for libraries to include (p5.js, three.js, etc.)

Would be nice to be able to control the script that ingests the left side content to render to the right side--this should make it dramatically easier to create interpreters and renderers for all sorts of other applications--for example being able to take the content and render it thru nomnoml or d2 (alternatives to mermaid) or run through a JSON validator. It would even allow fast implementations of other suggestions, like markdown rendering which is mentioned in this thread.
